### DISCUSSION TITLE
일대다 관계를 일급 컬렉션으로 매핑하기

### DISCUSSION THREAD

**Discussion by kmw2378** [KakaoFunding/back-end#33] [Ideas]
일대다 관계를 `List`로 관리하게 되면 `stream()` 중복 코드가 발생하게 됩니다.
e.g.  `Member`와 `Role` 이 일대다 관계
```java
public class Member {
    @OneToMany(mappedby = "member")
    private List<Role> roles;
   ....
}
```

```java
public class Role {
    private Long id;
    private RoleValue value;
    
    @ManyToOne
    @JoinColumn(name = "member_id")
    private Member member;
    
    ...
}
```

만약, 비지니스 레이어에서 특정 회원의 모든 `roleValue`를 받아온다면 아래와 같습니다.
```java
List<String> values = member.getRoles().stream()
    .map(role -> role.getValue().name()
    .toList();
```

이러한 로직이 다른 곳에도 사용된다면 중복 코드가 발생하게 되고 `List`에 대한 불변성 보장도 힘듭니다. 

이번 프로젝트에서 일대다관계를 이렇게 풀어보는건 어떨까 합니다!

[관련 링크](https://bgpark.tistory.com/156)

**Comment by hwan2-99** [comment]
이 부분에 대해선 제가 이전에 공부하다가 되는지 몰라서 포기한 부분인데 엔티티가 임베디드 타입으로 변경이 되는군요. .하지만 지금처럼 도메인이 많이 나누어져 있는 상황에서 모든걸 일급컬렉션을 만들기위해 클래스화 한다면 관리가 꽤 어려워질것 같기도 한데 그럼에도 불구하고 사용해야할만큼 이점이 있을까요??

**Reply by kmw2378** [reply]
말씀하신 것과 같은 트레이드 오프가 있습니다! 

그러나, 일대다 관계가 많을수록 관련 중복 코드를 줄일 수 있고 궁극적으로 캡슐화를 더 강하게 할 수 있어 좋다 생각합니다!

**Comment by YeaChan05** [comment]
확실히 적당한 추상화를 잘 사용해서 구현한다면 충분히 책임 분리가 가능할거 같네요
하지만 걱정되는 부분이 몇가지 있습니다
1. querydsl entity graph 탐색이 가능한지
2. 지연로딩이 적용 되는지
3. 과연 확장성과 개발비용을 합쳤을때 효율적일지
4. 더티 채킹이 적용 되는지

이러한 부분들을 모두 고려했을때 긍정적 효과를 거둘 수 있다고 한다면 충분히 사용해도 좋다고 생각합니다
하지만 현재 알고있는 지식 바탕에서도 예측 불가한 예외가 자주 발생하는데, jpa가 엄청 깊은 지식이 필요하다 보니 괜히 배보다 배꼽이 더 커지는 느낌이 드네요... 추후 성능 개선 및 리팩토링때 다시 논해보는게 어떨까요?

**Reply by YeaChan05** [reply]
추가적으로, 해당 내용은 영속 모델과 도메인 모델이 분리되었을때 더 적합해보인다고 생각됩니다!
